Library/Homebrew/cask/cmd.rb

```diff
@@ -180,24 +179,38 @@ def self.nice_listing(cask_list)
     end
 
     def process_options(*args)
-      exclude_regex = /^--#{Regexp.union(*Config::DEFAULT_DIRS.keys.map(&Regexp.public_method(:escape)))}=/
-
       non_options = []
 
       if idx = args.index("--")
         non_options += args.drop(idx)
         args = args.first(idx)
       end
 
+      exclude_regex = /^--#{Regexp.union(*Config::DEFAULT_DIRS.keys.map(&Regexp.public_method(:escape)))}=/
       cask_opts = Shellwords.shellsplit(ENV.fetch("HOMEBREW_CASK_OPTS", ""))
                             .reject { |arg| arg.match?(exclude_regex) }
 
       all_args = cask_opts + args
 
-      remaining = all_args.select do |arg|
-        !process_arguments([arg]).empty?
-      rescue OptionParser::InvalidOption, OptionParser::MissingArgument, OptionParser::AmbiguousOption
-        true
+      i = 0
+      remaining = []
+
+      while i < all_args.count
+        begin
+          arg = all_args[i]
+
+          remaining << arg unless process_arguments([arg]).empty?
+        rescue OptionParser::MissingArgument
```

Comment:
Would be good to try and use `cli/parser` here ASAP.

Reply:
I had a look and it seems like the `cli/parser` will need quite some refactoring in order to support cask subcommands, or rather multiple levels of flags: one for `brew`, `brew cask` and `brew cask <subcommand`.

Reply:
@reitermarkus Cool. Are you game to start some of that? If not: could you create a help wanted issue to discuss the steps as you see them that are required?